src/AutoRest.CSharp/Mgmt/Decorator/PropertyMatchDetection.cs

```diff
@@ -98,9 +98,14 @@ internal static bool DoesPropertyExistInParent(Type sourceType, MgmtObjectType t
 
             if (!parentDict.TryGetValue(childProperty.Declaration.Name, out parentProperty))
             {
-                if (childProperty.Declaration.Name.EndsWith("Type"))
+                // If exact property name match fails, try to match ResourceType with Type, match ManagedServiceIdentityType and SystemAssignedServiceIdentityType with Type or XXType.
+                if (childProperty.Declaration.Name.Equals("Type", StringComparison.Ordinal))
                 {
-                    parentProperty = parentDict.FirstOrDefault(p => p.Key.EndsWith("Type")).Value;
+                    parentProperty = parentDict.FirstOrDefault(p => p.Key.EndsWith("Type", StringComparison.Ordinal)).Value;
+                }
+                else if (childProperty.Declaration.Name.EndsWith("Type", StringComparison.Ordinal))
+                {
+                    parentProperty = parentDict.FirstOrDefault(p => p.Key.EndsWith("Type", StringComparison.Ordinal) && !p.Key.Equals("ResourceType", StringComparison.Ordinal)).Value;
```

Comment:
I have a few concerns about this piece of code (not from this PR)
This matching logic is kind of fragile, we are matching things if they ends with the word `Type`, would this matches something that are totally irrelevant? What if the class we are trying to compare has two different properties ending with the word `Type`?

Before this causes any issues, we might need to change this to compare the field name which this property serializes into

Reply:
The original code caused issue as I didn't expect the method was used separately in [superset match](https://github.com/Azure/autorest.csharp/blob/15c1e6cee6998c3c0d3088231804e007f5fa3797/src/AutoRest.CSharp/Mgmt/Decorator/InheritanceChooser.cs#L93https://github.com/Azure/autorest.csharp/blob/15c1e6cee6998c3c0d3088231804e007f5fa3797/src/AutoRest.CSharp/Mgmt/Decorator/InheritanceChooser.cs#L93). I thought it was only used in exact match and this fuzzy match relies on the checking of property count match and exact match of other properties. The change here removed the fuzzy match for superset match where the only property ending with Type is ResourceType. 

As for the field name match, we do not have that information for ReferenceTypes like ResourceData in the SDK code as we don't even have a serialization class for it. 

In general, I agree that the property match logic needs to be enhanced, it has too many patches now and could break with new corner cases. One potential improvement is that we should remove a property from parentDict after it's matched to avoid matching two different properties ending with the word Type with the same parent property as you mentioned.